src/marketplace.js

```diff
@@ -0,0 +1,377 @@
+     * @property {string} hash calculated automatically
+     * @property {string} signature applied automatically
+     */
+
+    /**
+     * Create/Update the vendor signature for a vendor-marketplace key
+     * 
+     * @param {string} vendorId the vendor's public key multihash
+     * @returns {Promise}
+     * @throws {Error} if the signature is not valid or the vendor does not exist
+     * @memberof Marketplace
+     */
+    async updateVendorSignature(vendorPubKeyMultihash, signature) {
+        const id = vendorPubKeyMultihash;
+        validateMultihash(id);
+        await this.start();
+        try {
+            const vendor = await this.getVendor(id);
+            // TODO: signature needs expiration date?
+            const PvmMultihash = vendor.vmPubKeyMultihash;
+            const valid = this.chluIpfs.instance.crypto.verifyMultihash(vendorPubKeyMultihash, PvmMultihash, signature);
+            if (valid) {
+                vendor.vSignature = signature;
+                await this.db.updateVendor(id, vendor);
+            } else {
+                throw new HttpError(400, 'Signature is not valid');
+            }
+        } catch (error) {
+            if (error instanceof HttpError) throw error;
+            throw new HttpError(500, 'An error has occurred: ' + error.message);
+        }
+    }
+
+    /**
+     * Create a Proof of Payment Request for a vendor.
+     * 
+     * @param {string} vendorId the vendor's public key multihash
+     * @param {object} [options={}]
+     * @param {string} options.item_id
+     * @param {string} options.invoice_id
+     * @param {string} options.customer_id
+     * @param {integer} options.created_at
+     * @param {integer} options.expires_at
+     * @param {string} options.currency_symbol
+     * @param {integer} options.amount
+     * @param {string} options.marketplace_url
+     * @param {string} options.marketplace_vendor_url
+     * @param {string} options.key_location
+     * @returns {Promise<PoPR>}
+     * @memberof Marketplace
+     */
+    async createPoPR(vendorId, options = {}) {
+        validateMultihash(vendorId);
+        await this.start();
+        try {
+            const vendor = await this.getVendor(vendorId);
+            const popr = {
+                item_id: options.item_id || 'N/A',
+                invoice_id: options.invoice_id || 'N/A',
+                customer_id: options.customer_id || 'N/A',
+                created_at: options.created_at || Date.now(),
+                expires_at: options.expires_at || 0,
+                currency_symbol: options.currency_symbol || 'N/A',
+                amount: options.amount || 0,
+                marketplace_url: this.marketplaceLocation.slice(0),
+                marketplace_vendor_url: '/ipfs/' + vendor.vPubKeyMultihash,
+                key_location: '/ipfs/' + vendor.vmPubKeyMultihash,
+                vendor_key_location: '/ipfs/' + vendor.vPubKeyMultihash,
+                vendor_signature: vendor.vSignature,
```

Comment:
So the popr field 13 from protocol's protobuf specs is now being split into two fields? vendor_signature and marketplace_signature?

Also, might be useful that the marketplace returns a protobuf, so that we don't have versioning issues later on. Can you issue this up for later?

Reply:
@kulpreet here's an issue about how we serialize signatures and pubkeys https://github.com/ChluNetwork/chlu-ipfs-support/issues/64

Reply:
about returning a protobuf from the marketplace, you mean it should serialize the PoPR before returning it instead of using JSON? Why would there be issues if we don't do this?

Reply:
It is largely to allow other apps to use this library. We can leave it as it is for now.